File: csharp/MarsRover/MarsRover/Rover.cs
namespace MarsRover;

public enum Direction { N, E, S, W }

public record Location(int X, int Y)
{
    public static Location Origin => new(0, 0);
}

public enum Command { F, B, L, R }

public record Rover(Direction Direction, Location Location)
{
    public static Rover Initial => new(Direction.N, Location.Origin);

    public Rover Execute(Command[] commands) =>
        commands.Aggregate(
            this,
            (r, c) => r.Execute(c));

    public Rover Execute(Command c) =>
        c switch
        {
            Command.L => TurnLeft(),
            Command.R => TurnRight(),
            Command.F => MoveForward(),
            Command.B => MoveBackward(),
            _ => throw new ArgumentOutOfRangeException(nameof(c), c, null)
        };

    private Rover TurnLeft() =>
        Direction switch
        {
            Direction.N => this with { Direction = Direction.W },
            Direction.E => this with { Direction = Direction.N },
            Direction.S => this with { Direction = Direction.E },
            Direction.W => this with { Direction = Direction.S },
            _ => throw new ArgumentOutOfRangeException()
        };

    private Rover TurnRight() =>
        Direction switch
        {
            Direction.N => this with { Direction = Direction.E },
            Direction.E => this with { Direction = Direction.S },
            Direction.S => this with { Direction = Direction.W },
            Direction.W => this with { Direction = Direction.N },
            _ => throw new ArgumentOutOfRangeException()
        };

    private Rover MoveForward()
    {
        var (x, y) = this.Location;
        return Direction switch
        {
            Direction.N => this with { Location = new Location(x + 1, y) },
            Direction.E => this with { Location = new Location(x, y + 1) },
            Direction.S => this with { Location = new Location(x - 1, y) },
            Direction.W => this with { Location = new Location(x, y - 1) },
            _ => throw new ArgumentOutOfRangeException()
        };
    }

    private Rover MoveBackward()
    {
        var (x, y) = this.Location;
        return Direction switch
        {
            Direction.N => this with { Location = new Location(x - 1, y) },
            Direction.E => this with { Location = new Location(x, y - 1) },
            Direction.S => this with { Location = new Location(x + 1, y) },
            Direction.W => this with { Location = new Location(x, y + 1) },
            _ => throw new ArgumentOutOfRangeException()
        };
    }
}

File: python/MarsRover/rover.py
from __future__ import annotations

from enum import Enum


class Direction(Enum):
    N = "North"
    E = "East"
    S = "South"
    W = "West"


class Command(Enum):
    F = "Forward"
    B = "Backward"
    L = "Left"
    R = "Right"


class Location:
    x: int
    y: int

    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

    def __eq__(self, other: Location) -> bool:
        return self.x == other.x and self.y == other.y

    def __str__(self):
        return "["+str(self.x)+", "+str(self.y)+"]"

    @staticmethod
    def origin():
        return Location(0, 0)


class Rover:
    direction: Direction
    location: Location

    def __init__(self, location: Location, direction: Direction):
        self.direction = direction
        self.location = location

    def __eq__(self, other: Rover) -> bool:
        return self.location == other.location and self.direction == other.direction

    def __str__(self):
        return "Rover at "+str(self.location) + " facing " + str(self.direction)

    @staticmethod
    def initial() -> Rover:
        return Rover(Direction.N, Location.origin())

    def execute(self, commands: list[Command]) -> Rover:
        for c in commands:
            self.execute_singular(c)
        return self

    def execute_singular(self, command: Command) -> Rover:
        match command:
            case Command.F: self.move_forward()
            case Command.B: self.move_backward()
            case Command.L: self.turn_left()
            case Command.R: self.turn_right()
        return self

    def move_forward(self) -> Rover:
        match self.direction:
            case Direction.N: self.location = Location(self.location.x+1, self.location.y)
            case Direction.E: self.location = Location(self.location.x, self.location.y+1)
            case Direction.S: self.location = Location(self.location.x-1, self.location.y)
            case Direction.W: self.location = Location(self.location.x, self.location.y-1)
        return self

    def move_backward(self) -> Rover:
        match self.direction:
            case Direction.N: self.location = Location(self.location.x-1, self.location.y)
            case Direction.E: self.location = Location(self.location.x, self.location.y-1)
            case Direction.S: self.location = Location(self.location.x+1, self.location.y)
            case Direction.W: self.location = Location(self.location.x, self.location.y+1)
        return self

    def turn_left(self) -> Rover:
        match self.direction:
            case Direction.N: self.direction = Direction.W
            case Direction.E: self.direction = Direction.N
            case Direction.S: self.direction = Direction.E
            case Direction.W: self.direction = Direction.S
        return self

    def turn_right(self) -> Rover:
        match self.direction:
            case Direction.N: self.direction = Direction.E
            case Direction.E: self.direction = Direction.S
            case Direction.S: self.direction = Direction.W
            case Direction.W: self.direction = Direction.N
        return self



